p4-16/psa/PSA.mdk

```diff
@@ -1497,6 +1499,59 @@ table.
 A PSA implementation must implement tables that have both a
 `psa_direct_counter` and `psa_direct_meter` property.
 
+### Table entry timeout notification {#sec-idle-timeout}
+
+PSA uses the `psa_idle_timeout` to enable a table implementation send
+notifications from the PSA device when a configurable time has passed since an
+entry was last matched. The property may take one of two values --
+`PSA_NO_TIMEOUT`, and `PSA_NOTIFY_CONTROL`. `PSA_NO_TIMEOUT` disables idle
+timeout support for the table. `PSA_NOTIFY_CONTROL` enables the notification. A
+PSA implementation will then generate an API for the control plane set
+time-to-live (TTL) values for table entries and if at any time during its
+lifetime, the table entry is not "hit" (&ie; not selected by any packet lookup)
+for a lapse of time greater or equal to its TTL, the device should generate a
+notification to the control plane.
+
+Example:
+```
+enum PSA_IdleTimeout_t {
+  PSA_NO_TIMEOUT,
+  PSA_NOTIFY_CONTROL
+}
+
+table t {
+  key = {};
+  actions = {};
+  psa_idle_timeout = PSA_NOTIFY_CONTROL;
+}
+```
+
+Restrictions on the TTL values and notifications:
+
+- It is likely that any hardware implementation will have a limited number of
+  bits to represent the values, and, since the values are programmed at
+  runtime, it is the responsability of the runtime (P4Runtime or other
+  controller software) to guarantee that the TTL values can be represented in
+  the device. This can be done by scaling the values to the number of bits
+  available on the platform, ensuring that the range of values between
+  different entries are representable. A PSA implementation should only enable
+  the programming of such tables, and return an error if the device does not
+  support the idle timeout at all.
+
+- If no value is programmed for a table entry, even though the table has
+  enabled the idle timeout property, the entry will not generate a
+  notification. The entry will participate in the aging process, with the
+  maximum timeout interval possible, i.e., its idle timeout counter will be set
+  to the maximum interval possible and the entry may be deleted when the
+  maximum interval elapses.
+
+- PSA does not require a timeout value for a default action entry. The reason
+  for not making this mandatory in the specification is that the default action
+  may not have an actual entry in the table and it is may be hit on the miss
+  table misses. Not sure what an TTL means for a miss path. The default action entry will not be aged out.
```

Comment:
nit: wrap the text?

Reply:
@antoninbas @cc10512 Would you be willing to specify that for tables with psa_idle_timeout equal to NOTIFY_CONTROL, that they _must never_ generate a notification message for the 'default entry'?  Even if an implementation actually implemented `default_action` using some wildcard ternary table entry with lowest priority, for example, the P4Runtime server software could avoid generating any timeout notification messages to the controller if that entry went unmatched for a long time.

Does the P4Runtime API allow the controller to specify a timeout duration for a table's default entry?  I don't think it needs to, personally, but am open to other views on that detail.

Reply:
P4Runtime does not make the distinction between the default entry & regular match entries in this case.
I feel like it does not harm to leave it as optional. I can add a note in the P4Runtime spec stating that idle timeout support for default entry is optional to keep things consistent. But I also do not see a use case for this specific feature, so I'm fine with prohibiting the use of idle timeout with default entries.

Reply:
I'd prefer to leave it to implementers as well. Even though it may seem counter to portability, there will be situations where one might want timeouts on default entries, and then it is just a matter of the target supporting it or not. Unfortunately it is not something that the compiler can detect, so runtime API will have to have the means of knowing whether setting the timeout succeeded.